### PR TITLE
Fixes #1111 : Added Forward-Forward Algorithm

### LINKED ISSUE #1111: 🚀 Feature request / I want to contribute an algorithm
## Is your feature request related to a problem? Please describe.

Currently, PyTorch/examples does not have an implementation of the forward forward algorithm.[forward forward algorithm.](https://arxiv.org/abs/2212.13345)  This algorithm is a new learning procedure for neural networks and has promising approach to training neural networks, it is also becoming popular, because it's written by father of deep learning aka Geoffrey Hinton, its inclusion in PyTorch/examples would make it more accessible to a wider community of researchers  practitioners, and I would like to contribute in it❤️, I've Implemented This algorithm in my local notebook in pure pytorch❤️.I am new so please let me know How can I contribute this algorithm in this repo.

Thanks,
Vivek

## Describe the solution

The solution is to implement/add the forward forward algorithm in PyTorch/examples. This would include writing the code for the algorithm, as well as any docs or tutorial addition to the existing codebase.

## Describe alternatives solution

[https://keras.io/examples/vision/forwardforward/](https://keras.io/examples/vision/forwardforward/)

### PR DESCRIPTION
**Pull request by viveks-codes**
this PR includes code for [Forward-Forward Algorithm](https://arxiv.org/abs/2212.13345),
as described in issue [#1111 ](https://github.com/pytorch/examples/issues/1111).

Thanks❤️,
Vivek V patel.

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -17,6 +17,9 @@ run_python_examples.sh @msaroufim
 ./reinforcement_learning/ @msaroufim 
 ./word_language_model/ @msaroufim
 
+# Forward Forward Examples
+./mnist_forward_forward/ @viveks-codes
```
Comment (msaroufim) [inline CODEOWNERS:21]: This is specific to people with write access to the repo so might be best to revert it. Happy to change this assumption if you end up being a regular contributor

File: mnist_forward_forward/main.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        ]
+    )
+    train_loader = DataLoader(
+        MNIST("./data/", train=True, download=True, transform=transform), **train_kwargs
+    )
+    test_loader = DataLoader(
+        MNIST("./data/", train=False, download=True, transform=transform), **test_kwargs
+    )
+    net = Net([784, 500, 500])
+    x, y = next(iter(train_loader))
```
Comment (msaroufim) [inline mnist_forward_forward/main.py:168]: just to clarify: this does a single step? Curious if this implementation actually converges

**Review by msaroufim — CHANGES_REQUESTED** [review]
I'm a bit concerned that this code borrows a bit too liberally from https://github.com/mohammadpz/pytorch_forward_forward/blob/main/main.py

I don't think it'd be fair to OP for me to merge this, at the very least we'd need to credit the OP and the MIT license. The arparser and docs are a nice addition

If @mohammadpz would like to upstream his work here for added visibility then it might make more sense to make further improvements to his work but if not then I don't believe I can merge this

**Comment by viveks-codes** [comment]
I appreciate your comments on my pull request. I recognise your concern for giving the original author the due credit and adhering to the MIT licence. I've sent an email to @mohammadpz, I've been in touch with the creator of the original code and asked for guidance on how to appropriately acknowledge their contribution. 

I will amend my pull request to include the correct attribution and licencing information as soon as I obtain the information I need. If you would like, I can also add documentation and an argparse, as you requested, to the code, among other changes.

I appreciate you giving this issue your time and attention. I'm eager to collaborate with you to merge this code into the PyTorch repository.

best,
Vivek

**Comment by mohammadpz** [comment]
Hi there!
Thank you for including me in this conversation, @msaroufim! I actually haven't received an email from @viveks-codes, or I may have missed it somehow.
Nonetheless, my code is licensed under MIT, so I have no issue with it being integrated into PyTorch as long as the license is updated to acknowledge the source of the original code.
Thanks!

**Comment by viveks-codes** [comment]
@mohammadpz 
Can you confirm that the email on your website is correct? Because I sent you an email about this a while ago, but anyway thank you very much!❤️

File: mnist_forward_forward/main.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            Lambda(lambda x: torch.flatten(x)),
+        ]
+    )
+    train_loader = DataLoader(
+        MNIST("./data/", train=True, download=True, transform=transform), **train_kwargs
+    )
+    test_loader = DataLoader(
+        MNIST("./data/", train=False, download=True, transform=transform), **test_kwargs
+    )
+    net = Net([784, 500, 500])
```
Comment (viveks-codes) [inline mnist_forward_forward/main.py:171]: Well, as you stated, @msaroufim, the algorithm does not converge on a single epoch because it assigns a single label to every instance. As a result, the train and test errors are approximately 90% accurate all the time (for example, as you may be aware with MNIST if you assign every x a single label, you are correct 90% of the time, always).

Because of this, the train and test errors for a single epoch are both about 90.

however,
Default epochs, are 1000, and mistakes are:

test error: 0.07350003719329834 
train error: 0.07302004098892212

**Review by msaroufim — APPROVED** [review]
